keps/sig-node/4742-node-topology-downward-api/README.md

```diff
@@ -0,0 +1,409 @@
+adding topology labels that match those of the target Node. The Binding REST implementation will be updated to copy these labels
+from the `pods/binding` subresource to the assigned Pod's labels.
+
+Using the downward API to retrieve topology information will look similar to the following:
+```
+apiVersion: v1
+kind: Pod
+metadata:
+  generateName: downwardapi-example-
+spec:
+  containers:
+    - name: container
+      image: registry.k8s.io/busybox
+      command: ["sh", "-c"]
+      args:
+      - while true; do
+          if [[ -e /etc/podinfo/zone ]]; then
+            echo -en '\n\n'; cat /etc/podinfo/zone; fi;
+          sleep 5;
+        done;
+      volumeMounts:
+        - name: podinfo
+          mountPath: /etc/podinfo
+  volumes:
+    - name: podinfo
+      downwardAPI:
+        items:
+          - path: "zone"
+            fieldRef:
+              fieldPath: metadata.labels['topology.kubernetes.io/zone']
+```
+
+### User Stories
+
+* As an ML engineer, I want to optimize GPU-to-GPU communication during training which requires topology-awareness in my training code.
+* As a database developer, I want to leverage topology information to improve fault tolerance of sharded databases on Kubernetes.
+* As a developer of a Kubernetes CNI plugin, I want to pass topology information down to the CNI plugin to optimize data paths for container networks.
+
+### Notes/Constraints/Caveats (Optional)
+
+### Risks and Mitigations
+
+* Scope creep. Allowing additional node information or node label info could
+create security issues. This is mitigated by limiting the node labels
+to strictly those that are standardized through KEP-1659.
+
+* Exposing sensitive data as node labels to pods. This is mitigated by ensuring
+standard topology labels are available to Pods.
+
+* Stale data. Information obtained through node labels is like information
+attained through a configmap or secret mounted to a pod, being passed on
+creation but not guaranteed to be immutable and thus should be treated as so.
+
+
+## Design Details
+
+* A built-in Kubernetes admission plugin, `PodTopologyLabels` will be introduced in kube-apiserver
+* The `PodTopologyLabels` admission plugin is responsible for mutating `pods/binding` subresource, adding topology labels matching the target Node.
+* `PodTopologyLabels` admission will not override any existing topology labels on Pods.
+* A feature gate, `PodTopologyLabelsAdmission` will be introduced in v1.33. Alpha and disabled by default.
+The `PodTopologyLabels` admission plugin can only be set when this feature gate is enabled.
+* The Binding REST implementation will be updated to copy all labels from `pods/binding` subresource into Pods.
+At this point we do not remove labels, it is assumed that any labels in Binding are safe to expose via Downward API.
+* For exposing additional node labels, at the discretion of the cluster admin, a mutating admission webhook can be used to mutate labels of the `pods/binding` subresource.
```

Comment:
so all labels from pods/binding will be available? Just checking

Reply:
Not just available, but copied to the Pod object. This is under the assumption that nothing is setting labels on pods/binding today and this would be a first use-case for it.

Reply:
@thockin @enj @liggitt are we happy with this behaviour? To summarise:

1) the admission plugin will copy `topology.k8s.io/{zone,region}` + `kubernetes.io/hostname` from the Node object to the Binding during `pods/binding` admission
2) the Binding registry will indiscriminately copy all labels on the `Binding` object to the Pod, regardless of their prefix

This is nice and extensible as it does mean any scheduler (or mutating admission webhook/policy) can mutate `pods/binding` to inject their own arbitrary organisation specific labels (e.g. `example.com/datahall`), but on the other hand could be confusing if we overwrite existing labels on the Pod...

Reply:
https://github.com/kubernetes/enhancements/pull/5146#discussion_r1953511000 feels relevant.